server/base/src/main/java/org/apache/accumulo/server/util/UpgradePreparationUtil.java

```diff
@@ -0,0 +1,155 @@
+
+@AutoService(KeywordExecutable.class)
+public class UpgradePreparationUtil implements KeywordExecutable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(UpgradePreparationUtil.class);
+
+  static class Opts extends Help {
+    @Parameter(names = "--prepare-for-upgrade",
+        description = "prepare an older version instance for an upgrade to a newer non-bugfix release."
+            + " This command should be run using the older version of software after the instance is shut down.")
+    boolean postShutdownUpgradeCheck = false;
+
+    @Parameter(names = "--force", description = "allow --prepare-for-upgrade to run again")
+    boolean force = false;
+  }
+
+  @Override
+  public String keyword() {
+    return "upgrade";
+  }
+
+  @Override
+  public String description() {
+    return "utility used to prepare an instance for a minor or major version upgrade";
+  }
+
+  @Override
+  public void execute(String[] args) throws Exception {
+    Opts opts = new Opts();
+    opts.parseArgs(keyword(), args);
+
+    if (!opts.postShutdownUpgradeCheck) {
+      new JCommander(opts).usage();
+      return;
+    }
+
+    var siteConf = SiteConfiguration.auto();
+    // Login as the server on secure HDFS
+    if (siteConf.getBoolean(Property.INSTANCE_RPC_SASL_ENABLED)) {
+      SecurityUtil.serverLogin(siteConf);
+    }
+
+    String volDir = VolumeConfiguration.getVolumeUris(siteConf).iterator().next();
+    Path instanceDir = new Path(volDir, "instance_id");
+    InstanceId iid = VolumeManager.getInstanceIDFromHdfs(instanceDir, new Configuration());
+    ZooReaderWriter zoo = new ZooReaderWriter(siteConf);
+
+    if (opts.postShutdownUpgradeCheck) {
+      final String zUpgradepath = Constants.ZROOT + "/" + iid + Constants.ZPREPARE_FOR_UPGRADE;
+      try {
+        if (zoo.exists(zUpgradepath)) {
+          if (!opts.force) {
+            throw new IllegalStateException(
+                "'accumulo upgrade --prepare-for-upgrade' must have already been run."
+                    + " To run again use 'accumulo upgrade --prepare-for-upgrade --force'");
+          } else {
+            zoo.delete(zUpgradepath);
+          }
+        }
+      } catch (KeeperException | InterruptedException e) {
+        throw new IllegalStateException("Error creating or checking for " + zUpgradepath
+            + " node in zookeeper: " + e.getMessage(), e);
+      }
+
+      LOG.info("Upgrade specified, validating that Manager is stopped");
+      final ServiceLockPath mgrPath =
+          ServiceLock.path(Constants.ZROOT + "/" + iid + Constants.ZMANAGER_LOCK);
+      try {
+        if (ServiceLock.getLockData(zoo.getZooKeeper(), mgrPath) != null) {
+          throw new IllegalStateException(
+              "Manager is running, shut it down and retry this operation");
+        }
+      } catch (KeeperException | InterruptedException e) {
+        throw new IllegalStateException("Error trying to determine if Manager lock is held", e);
```

Comment:
Can we add some primitive to ZooZap we can just call out here? Like `--managers --test`? Or maybe we'd have to wait until 4.0 for stuff like that?

Reply:
I think we should add that to 4.0 vs adding a feature into ZooZap in a bugfix release.

Reply:
I don't think we 'll need to modify ZooZap in 4.0. There is the ServerPaths object accessible directly from the ServerContext.